posts/bayesian-model-comparison/index.qmd

```diff
@@ -0,0 +1,470 @@
+subtitle: 'Model selection via a Bayesian correlated t-test.'
+date: 2022/11/11
+image: thumb.png
+categories:
+  - ML
+  - Theory
+  - Model evaluation
+  - Bayes
+  - Bayesian estimation
+  - Cross-validation
+bibliography: ../references.bib
+jupyter:
+  jupytext:
+    text_representation:
+      extension: .qmd
+      format_name: quarto
+      format_version: '1.0'
+      jupytext_version: 1.13.7
+  kernelspec:
+    display_name: Python 3.6.15 ('fgm')
+    language: python
+    name: python3
+---
+
+```{python}
+#| eval: false
+#| echo: false
+!pip install -r requirements.txt
+```
+
+Robust model evaluation should be second nature for those of us that work with data and 
+predictive models. 
+When determining whether one model is better than another, there are many techniques one can use.
+Very common ones include *bootstrapping* and *cross-validation*. 
+
+First and foremost however, one must make a decision as to what "better" means.
+Better in what way? 
+More accurate? 
+Faster? 
+More (budget) efficient? 
+We will come to this point again later.
+For now, let's assume that better means "obtains a higher F1 score".
+F1 is just an example, and we could consider any other score. 
+Once we have decided to say that the model with highest F1 score is better, the more serious question arises: how do we know whether a model is indeed better? 
+Can we just take an average score across a test set and compare the numbers, i.e. plain-old pointwise comparison? 
+I think many will get a sense of discomfort just uttering that. 
+We would like something more robust and which, ideally, takes uncertainty into account. 
+
+We ideally want to have a probability distribution which informs us, in some way, what the probability of one model being better than the other is. 
+That is, we want something like $P(M_1 > M_2 | D)$, where $M_1 > M_2$ denotes that model $M_1$ is better than model $M_2$ and $D$ is the data used for the comparison.
+In fact, we may want to do more than that. 
+Ideally, we would want a distribution specifying what the probability of one model being better than another is for each given performance difference. 
+That is, we would like a distribution over the true performance difference of the two algorithms.
+What a great motive to enter the wonderful world of *Bayesian modelling*[^1]. 
+
+[^1]: See, e.g. @kruske2013best, for a nice, rapid tour of Bayesian modelling for group comparisons.
+
+## Problem statement
+
+Assume that we are interested in comparing the performance (accuracy, recall, precision, $\text{F}_{\beta}$, etc.) of two predictors.
+If using cross-validation, @bouckaert2003choosing recommends making this more robust (i.e. return more stable statistics) by using *repeated cross-validation*; that is performing evaluation via $m$ repetitions of $k$-fold cross-validation. 
+Both classifiers will be trained on the same data and tested on the same data and we can therefore collect the *differences of their performances* 
```

Comment:
> Both classifiers will be ~trained on the same data and~ tested on the same data and we can therefore collect the *differences of their performances*

Feel like training is a distraction.

Reply:
Hm, I'm not sure. It's important that they also be trained on the same sets as otherwise the models could be generated from different datasets. I could put it in parenthesis?

Reply:
interesting. would this be a problem for computing delta?

i see below training and testing become intertwined due to cross-validation, hence rho. but not sure i see it here.

Reply:
The rho correlation is the correction because of the samples not being independent (Samples being the scores, or difference of scores also) however we also need to have the same train set and same test set for the battle to be fair. Even in the case of plain train-test split it would be unfair to have one more trained on partition A and tested on C and another trained on B and tested on C.

Reply:
agree about like-for-like, not sure it's necessary tho. and indeed we dont say so.

think we agree, so happy to leave.

Reply:
I'll leave it then.